sklearn/model_selection/_validation.py

```diff
@@ -977,19 +1076,61 @@ def validation_curve(estimator, X, y, param_name, param_range, groups=None,
     X, y, groups = indexable(X, y, groups)
 
     cv = check_cv(cv, y, classifier=is_classifier(estimator))
-    scorer = check_scoring(estimator, scoring=scoring)
+    n_splits = cv.get_n_splits(X, y, groups)
+    n_params = len(param_range)
 
+    scorers, is_multimetric = _check_multimetric_scoring(estimator,
+                                                         scoring=scoring)
     parallel = Parallel(n_jobs=n_jobs, pre_dispatch=pre_dispatch,
                         verbose=verbose)
     out = parallel(delayed(_fit_and_score)(
-        estimator, X, y, scorer, train, test, verbose,
+        estimator, X, y, scorers, train, test, verbose,
         parameters={param_name: v}, fit_params=None, return_train_score=True)
         # NOTE do not change order of iteration to allow one time cv splitters
         for train, test in cv.split(X, y, groups) for v in param_range)
+    out = list(zip(*out))
+
+    train_scores = _aggregate_score_dicts(out[0], shape=(n_splits, n_params),
+                                          transpose=True)
+    test_scores = _aggregate_score_dicts(out[1], shape=(n_splits, n_params),
+                                         transpose=True)
+
+    if not is_multimetric:
+        # Return the score array directly
+        train_scores = list(train_scores.values())[0]
+        test_scores = list(test_scores.values())[0]
+
+    return train_scores, test_scores
 
-    out = np.asarray(out)
-    n_params = len(param_range)
-    n_cv_folds = out.shape[0] // n_params
-    out = out.reshape(n_cv_folds, n_params, 2).transpose((2, 1, 0))
 
-    return out[0], out[1]
+def _aggregate_score_dicts(scores, shape=None, transpose=False):
+    """Aggregate the list of dict to dict of np ndarray
+
+    The aggregated output of _fit_and_score will be a list of dict
+    # of form [{'prec': 0.1, 'acc':1.0}, {'prec': 0.1, 'acc':1.0}, ...]
+    # Convert it to a dict of array {'prec': np.array([0.1 ...]), ...}
+
+
+    If shape is set, the individual numpy arrays are reshaped accordingly
+
+    If transpose is set, the individual arrays are transposed after reshaping
+    This is useful if the the scores are not ordered first by cv splits.
+    """
+    if shape is None:
+        shape = (len(scores), )
+    else:
+        if np.product(shape) != len(scores):
+            raise ValueError("Shape must conform to the size of the scores")
+
+    np_empty = partial(np.empty, shape=(len(scores),), dtype=np.float64)
+    scores_arr = defaultdict(np_empty)
```

Comment:
I don't see the benefit of using defaultdict here.

I'd even just do something like:
```
for k in scores[0]:
    k_scores = np.asarray([s[k] for s in scores])
    # or maybe k_scores = np.fromiter((s[k] for s in scores), count=len(scores), dtype=np.float64)
    k_scores = k_scores.reshape(shape)
    if transpose:
        k_scores = k_score.T
    out[k] = k_scores
```

I don't think you should so a dtype conversion.

I'd add an assertion that the length of all elements of scores is equal afterward.